community/cypher/cypher/src/test/java/org/neo4j/cypher/CreateIndexStressIT.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2002-2017 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher;
+
+import org.junit.Rule;
+import org.junit.Test;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.neo4j.graphdb.factory.GraphDatabaseBuilder;
+import org.neo4j.graphdb.factory.GraphDatabaseSettings;
+import org.neo4j.test.rule.ImpermanentDatabaseRule;
+
+import static org.junit.Assert.assertFalse;
+
+public class CreateIndexStressIT
+{
+    private static final int NUM_PROPS = 400;
+    private final AtomicBoolean hasFailed = new AtomicBoolean( false );
+
+    @Rule
+    public ImpermanentDatabaseRule db = new ImpermanentDatabaseRule()
+    {
+        @Override
+        protected void configure( GraphDatabaseBuilder builder )
+        {
+            builder.setConfig( GraphDatabaseSettings.query_cache_size, "0" );
+        }
+    };
+
+    private final ExecutorService executorService = Executors.newFixedThreadPool( 10 );
+
+    @Test
+    public void shouldHandleConcurrentIndexCreationAndUsage() throws InterruptedException
+    {
+        // Given
+        HashMap<String,Object> params = new HashMap<>();
+        params.put( "param", NUM_PROPS );
+        db.execute( "FOREACH(x in range(0,$param) | CREATE (:A {prop:x})) ", params );
+        db.execute( "CREATE INDEX ON :A(prop) " );
```

Comment:
yes when we do planning we first check if there is an index and then check if there is a unique index,  https://github.com/neo4j/neo4j/blob/3.2/community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/planner/logical/steps/indexScanLeafPlanner.scala#L105. Since we now have changed the code not to use the `unique` variant but always use the normal index and filter lazily we need to explicitly wait for the index to come online.